rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,608 @@
+    bottom = 0
+    top = len(routing_table)
+    pos = (top - bottom) // 2
+
+    pg = gg(routing_table[pos])
+    while pg != generality and bottom < pos < top:
+        if pg < generality:
+            bottom = pos  # Move up
+        else:  # pg > generality
+            top = pos  # Move down
+
+        # Compute a new position
+        pos = bottom + (top - bottom) // 2
+        pg = gg(routing_table[pos])
+
+    while (pos < len(routing_table) and
+           gg(routing_table[pos]) <= generality):
+        pos += 1
+
+    return pos
+
+
+class Merge(namedtuple("Merge", ["routing_table", "entries", "key", "mask",
+                                 "generality", "goodness",
+                                 "insertion_index"])):
+    """Represents a potential merge of routing table entries.
+    """
+    def __new__(cls, routing_table, entries=set()):
+        # Generate the new key and mask
+        any_ones = 0x00000000  # Wherever there is a 1 in *any* of the keys
+        all_ones = 0xffffffff  # ... 1 in *all* of the keys
+        all_set = 0xffffffff  # ... 1 in *all* of the masks
+
+        for i in entries:
+            # Get the entry
+            entry = routing_table[i]
+
+            # Update the values
+            any_ones |= entry.key
+            all_ones &= entry.key
+            all_set &= entry.mask
+
+        # Compute the new mask, key and generality
+        any_zeros_or_xs = ~all_ones
+        mask = all_set & (any_ones ^ any_zeros_or_xs)
+        key = all_ones & mask
+        generality = get_generality(key, mask)
+        insertion_index = get_insertion_index(routing_table, generality)
+
+        # Compute the goodness of the merge
+        goodness = len(entries) - 1
+
+        return super(Merge, cls).__new__(
+            cls, routing_table, frozenset(entries), key, mask,
+            generality, goodness, insertion_index
+        )
+
+    def apply(self, aliases):
+        """Apply the merge to the routing table it is defined against and get a
+        new routing table and alias dictionary.
+
+        Returns
+        -------
+        [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+            A new routing table which may be minimised further.
+        {(key, mask): {(key, mask), ...}}
+            A new dictionary mapping keys and masks from the new routing table
+            to sets of keys and masks that they represent.
+        """
+        # Generate a new empty table of the correct size
+        routing_entries = list(self.routing_table)
+        aliases = dict(aliases)
+
+        # Add the new entry
+        new_entry = RoutingTableEntry(
+            route=routing_entries[next(iter(self.entries))].route,
+            key=self.key, mask=self.mask
+        )
+
+        routing_entries.insert(self.insertion_index, new_entry)
+        aliases[(self.key, self.mask)] = our_aliases = set([])
+
+        # Remove all the old entries
+        offset = 0
+        for i in sorted(self.entries):
```

Comment:
As noted in old version, an explicit note that the new entry insertion is guaranteed to be after all removed entries (and thus can't make their indexes wrong) would be a welcome reassurance.